### PR TITLE
[cxx] [glib] [coop] Auto void* to non-void*, int vs. enum, and other glib fixes.

### PR DESCRIPTION
**Pull request by jaykrell**
Mainly this deals with a large number of void* to non-void* conversions w/o line damage all around.
 They can be fixed individually and that was done and working but was not popular.
It also provides infrastructure for int vs. enum problem, and uses it some. This  is a much smaller problem in terms of line damage but this drives churn  further down.
And void*/non-void* is addressed in handle.h.

Basically, apply `g_cast` liberally around compilation errors.

### CODE REVIEW

File: mono/eglib/gunicode.c [hunk 1]
```diff
@@ -88,7 +88,7 @@ g_unichar_type (gunichar c)
 	/* since the argument is UTF-16, we cannot check beyond FFFF */
 
 	/* It should match any of above */
-	return 0;
+	return g_cast (0); // G_UNICODE_CONTROL
```
Comment (lambdageek) [inline mono/eglib/gunicode.c]: This seems like an abuse of `g_cast`.  You have to write something here and all the types are known, so just write the explicit cast.

File: mono/eglib/glib.h [hunk 2]
```diff
@@ -1148,4 +1408,24 @@ glong     g_utf8_pointer_to_offset (const gchar *str, const gchar *pos);
...
+#undef g_memdup
+#undef g_hash_table_lookup
+#define g_malloc(x) (g_cast (monoeg_malloc (x)))
+#define g_realloc(obj, size) (g_cast (monoeg_realloc ((obj), (size))))
+#define g_malloc0(x) (g_cast (monoeg_malloc0 (x)))
+#define g_calloc(x, y) (g_cast (monoeg_g_calloc ((x), (y))))
+#define g_try_malloc(x) (g_cast (monoeg_try_malloc (x)))
+#define g_try_realloc(obj, size) (g_cast (monoeg_try_realloc ((obj), (size))))
+#define g_memdup(mem, size) (g_cast (monoeg_g_memdup ((mem), (size))))
+#define g_hash_table_lookup(hash, key) (g_cast (monoeg_g_hash_table_lookup ((hash), (key))))
```
Comment (jaykrell) [inline mono/eglib/glib.h]: See for example

https://github.com/mono/mono/pull/10019/files#diff-98cf043c47e9ac369f935d88ea2c0e5dR517

https://github.com/mono/mono/pull/10019/files#diff-34ddcfc99bebbe2859c2bafcd20a2ac4R1320

for the slightly larger use.

File: mono/eglib/glib.h [hunk 3]
```diff
@@ -329,7 +571,13 @@ typedef void     (*GFreeFunc)      (gpointer       data);
  */
 typedef struct _GSList GSList;
 struct _GSList {
+#ifdef __cplusplus
+	// Allow for conversion of list->data to anything, in context.
+	// This saves adding casts to turn valid C into valid C++.
+	g_cast_t<gpointer> data;
```
Comment (lambdageek) [inline mono/eglib/glib.h]: This makes `_GSList` not a POD anymore.  Is this where the non-`explicit` default constructor for `g_cast_t<>` is used?

File: mono/eglib/gqsort.c [hunk 4]
```diff
@@ -40,14 +40,14 @@ typedef struct _QSortStack {
 	size_t count;
 } QSortStack;
 
-#define QSORT_PUSH(sp, a, c) (sp->array = a, sp->count = c, sp++)
+#define QSORT_PUSH(sp, a, c) (sp->array = g_cast (a), sp->count = c, sp++)
```
Comment (lambdageek) [inline mono/eglib/gqsort.c]: the result type is known, can just use a normal explicit cast to `(char*)`

File: mono/eglib/gqsort.c [hunk 5]
```diff
@@ -40,14 +40,14 @@ typedef struct _QSortStack {
...
 #define QSORT_POP(sp, a, c) (sp--, a = sp->array, c = sp->count)
 
 #define SWAPTYPE(TYPE, a, b) {              \
-	long __n = size / sizeof (TYPE);    \
-	register TYPE *__a = (TYPE *) (a);  \
-	register TYPE *__b = (TYPE *) (b);  \
-	register TYPE t;                    \
+	gssize __n = size / sizeof (TYPE);  \
+	TYPE *__a = (TYPE *) (a);           \
```
Comment (lambdageek) [inline mono/eglib/gqsort.c]: What's wrong with `register`?

File: mono/eglib/gqsort.c [hunk 6]
```diff
@@ -69,7 +69,7 @@ void
 g_qsort_with_data (gpointer base, size_t nmemb, size_t size, GCompareDataFunc compare, gpointer user_data)
 {
 	QSortStack stack[STACK_SIZE], *sp;
-	register char *i, *k, *mid;
+	char *i, *k, *mid;
```
Comment (lambdageek) [inline mono/eglib/gqsort.c]: Is removing `register` motivated by the C++ project or unrelated?

File: mono/eglib/gunicode.c [hunk 7]
```diff
@@ -62,7 +62,7 @@ g_unichar_type (gunichar c)
 			continue;
 		if (unicode_category_ranges [i].end <= cp)
 			continue;
-		return unicode_category [i] [cp - unicode_category_ranges [i].start];
+		return g_cast (unicode_category [i] [cp - unicode_category_ranges [i].start]);
```
Comment (lambdageek) [inline mono/eglib/gunicode.c]: The result type is known - just use an explicit cast to `(GUnicodeType)`

File: mono/eglib/glib.h [hunk 8]
```diff
@@ -54,6 +54,248 @@
...
+struct g_cast_t
+{
+private:
+	TFrom x;
+public:
+	explicit g_cast_t (TFrom y) : x(y)
+	{
+	}
+
+	g_cast_t () // Not really intended but has a user.
```
Comment (lambdageek) [inline mono/eglib/glib.h]: If I understand what this is for (to add `g_cast_t<>` data members to previously C-compatible structs), I think we're better off without it.

File: mono/eglib/glib.h [hunk 9]
```diff
@@ -54,6 +54,248 @@
...
+	bool operator >= (T y)
+	{
+		return x >= (TFrom)y;
+	}
+};
+
+// Member function compares g_cast_t to other, global compares other to g_cast_t,
+// i.e. which parameter is first vs. second.
+template <typename T1, typename T2>
+inline bool operator == (T1 x, g_cast_t<T2> y) // e.g. compare to NULL
```
Comment (lambdageek) [inline mono/eglib/glib.h]: First: `x` is `private`, so these all need to be `friend`s.

Second: this is a really really bad idea.

You're essentially throwing away all type safety (because `T1` and `T2` may have nothing to do with each other) so this will mask a ton of unintended bugs:

```cplusplus
int* f (void)
{
  return 0;
}

int
main ()
{
  g_cast_t<int*> z = g_cast ((int*)0);
  std::printf ("%d\n", f () == z);
  std::printf ("%d\n", f == z);
}
```

So in this case I accidentally wrote `f` instead of `f()` on the last line, and the answer changes from always true to always false without a warning (even with `-Wall -Wextra`).

File: mono/eglib/glib.h (hunk 3 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: This is not where it is used. There is a local variable out in the tree. I tried other fixed.

File: mono/eglib/gqsort.c (hunk 4 above)
Comment (jaykrell) [reply mono/eglib/gqsort.c]: g_cast becomes its own philosophy: no change to the C.

File: mono/eglib/glib.h (hunk 9 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: ‘Private’ was a very late addition. Probably a mistake. Let’s just remove it.

File: mono/eglib/gqsort.c (hunk 4 above)
Comment (jaykrell) [reply mono/eglib/gqsort.c]: But ok either way.

File: mono/eglib/glib.h (hunk 3 above)
Comment (lambdageek) [reply mono/eglib/glib.h]: I'm confused/surprised how there can be a local variable of `g_cast_t<>` type. It's a brand new type.

File: mono/eglib/gqsort.c (hunk 5 above)
Comment (jaykrell) [reply mono/eglib/gqsort.c]: ‘Register’ is a C++ warning at least, maybe error now or in future.

File: mono/eglib/gqsort.c (hunk 6 above)
Comment (jaykrell) [reply mono/eglib/gqsort.c]: C++ warning due to deprecation.

File: mono/eglib/glib.h (hunk 3 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: Local of ‘GSList’ or ‘GList’.

File: mono/eglib/glib.h (hunk 9 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: Type safety is not a goal here yet. Limiting line damage and making the code valid C & C++ & with same meaning.

But if you want TFrom only void* (or MonoObject*? Or any pointer?) and TTo only a pointer, then yes, agreed, I can try.

File: mono/eglib/glib.h (hunk 8 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: It avoids casts in uses of lists. For actual line damage reduction not to provide a well designed type. Same thing with most of the operators, including both orders. They might not all be used, but quite a few are. Including ordering, not just equality.

File: mono/eglib/gunicode.c (hunk 1 above)
Comment (jaykrell) [reply mono/eglib/gunicode.c]: “No change to C.” Useful approach?

File: mono/eglib/glib.h (hunk 9 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: But this also can help with function pointers fyi. But maybe fix those otherwise.

**Comment by jaykrell** [comment]
Higher level obvious question: add casts everywhere? It is tractable. I had it all done for x86/amd64/arm/arm64. Lots of green CI lanes.

File: mono/eglib/gqsort.c (hunk 4 above)
Comment (lambdageek) [reply mono/eglib/gqsort.c]: if we really want no change to C, we should do:
```
#if defined (__cplusplus)
#define MONO_CXX_CAST(T,X) ((T)(X))
#else
#define MONO_CXX_CAST(T,X) (X)
#endif
```

Using `g_cast` in cases where we can use a simpler workaround seems unnecessary.

**Comment by lambdageek** [comment]
@jaykrell The goal of `g_cast_t<>` should not be to eliminate "line damage", IMO.  It should be to solve a very specific problem that we have: there are a few places where the existing C code casts from `void*` to some non-`void*` type that isn't written down.  (ie the return type in the `HANDLE_FUNCTION_RETURN_OBJ` macro).  We should only use `g_cast_t<>` when we have no other choice.  It shouldn't have any operators or default- or copy-constructors.  It should solve one single problem for which we have no other workaround.

File: mono/eglib/glib.h (hunk 3 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: POD-ness should not affect layout. Perhaps ABI of pass/return-by-value, which should be rare/zero. We can cast in list uses instead. It isn’t as much as allocators. Also C/C++ should be pretty rare, that traffic in glib, if we convert utils/metadata/sgen/mini, as I have.

File: mono/eglib/gqsort.c (hunk 4 above)
Comment (jaykrell) [reply mono/eglib/gqsort.c]: ‘g_cast’ is also more mechanical consisent easier to read write. And most of all, easy to find later & improve. Explicit casts will be harder to find. So maybe a separate macro as you show? Some of these probably ok indefinitely but some (not necessarily here) less so. Let me fix enum and sort and we can look at others later.

**Comment by jaykrell** [comment]
There is workarounds for all of it, and a matter of degree. Auto is good for handle, and casts everywhere else, imho. 

But the casts drew objection, esp. for allocations.

I also think we could view this as temporary. Agree to cast in the long term? It is one solid tedious day & I already did it, but it is noisy, & we are not (yet) commited to C++.

**Comment by jaykrell** [comment]
I can try also to address the cases that need operators, but you might be surprised — comparison to ‘NULL’ gets here, and its type was ‘long’. That was why I first added operator== and !=. After I added them, I have no record of number of later uses.

**Comment by jaykrell** [comment]
The goal really was exactly line damage reduction. For handle, auto is good trivial C++11, or there are other options. Probably everywhere else knows the type and a local cast is easy, albeit very frequent.

**Comment by jaykrell** [comment]
Ok. I'm down to:

```
#define g_cast monoeg_g_cast // in case not inlined (see eglib-remap.h)

// g_cast converts void* to T*.
// e.g. #define malloc(x) (g_cast (malloc (x)))
// FIXME It used to do more. Rename?

struct g_cast
{
private:
	void * const x;
public:
	explicit g_cast (void *y) : x(y) { }
	g_cast (g_cast&& y) : x(y.x) { }
	g_cast () = delete;
	g_cast (const g_cast& y) = delete;

	template <typename TTo>
	operator TTo* () const
	{
		return (TTo*)x;
	}
};

#else

// FIXME? Parens are omitted to preserve prior meaning.
#define g_cast(x) x

#endif
```

The rvalue constructor is used by two places using ternary operator.
This can be fixed if needed, by inserting casts or not using ternary. It is probably ok.
List traversal, TLS, various trampolines, hashtable lookup, I all changed back to casts.
Leaving this just for "allocators", more than is in this PR.

Will update this PR later.

**Comment by jaykrell** [comment]
See https://github.com/mono/mono/commit/d0c43ac207358e6faf451a0c56b39cbe340f6ebe
for an idea of what line damage this traded for.

File: mono/eglib/gqsort.c (hunk 5 above)
Comment (jaykrell) [reply mono/eglib/gqsort.c]: `long` there was insidious -- having the same-ish meaning on all but one platform -- win64.

File: mono/eglib/gqsort.c (hunk 4 above)
Comment (jaykrell) [reply mono/eglib/gqsort.c]: I'm not against `MONO_CXX_CAST` but really we can probably add casts in C.

**Comment by jaykrell** [comment]
The level of type unsafety was pretty deliberate. You request it by saying “cast” or by using a list of void*. To shoehorn/bootstrap C into C++. One might argue though that was an opening for worse, and I used it as such, from any to any. It now (soon) narrower, from void* to pointer so I think will be more satisfactory.

File: mono/eglib/glist.c [hunk 10]
```diff
@@ -306,7 +306,7 @@ gpointer
 g_list_nth_data (GList *list, guint n)
 {
 	GList *node = g_list_nth (list, n);
-	return node ? node->data : NULL;
+	return node ? node->data : (gpointer)NULL;
```
Comment (jaykrell) [inline mono/eglib/glist.c]: Leftover, undo.

File: mono/eglib/gslist.c [hunk 11]
```diff
@@ -347,7 +347,7 @@ gpointer
 g_slist_nth_data (GSList *list, guint n)
 {
 	GSList *node = g_slist_nth (list, n);
-	return node ? node->data : NULL;
+	return node ? node->data : (gpointer)NULL;
```
Comment (jaykrell) [inline mono/eglib/gslist.c]: Leftover from earlier, undo.

File: mono/mini/mini-amd64.c [hunk 12]
```diff
@@ -7562,7 +7562,7 @@ mono_arch_emit_exceptions (MonoCompile *cfg)
 #endif /* DISABLE_JIT */
 
 void*
-mono_arch_instrument_prolog (MonoCompile *cfg, void *func, void *p, gboolean enable_arguments)
+(mono_arch_instrument_prolog) (MonoCompile *cfg, void *func, void *p, gboolean enable_arguments)
```
Comment (jaykrell) [inline mono/mini/mini-amd64.c]: Stretching the definition of "allocator", perhaps, but it a void*-returning function that users assign result to non-void* (guint8*). We have a mix of void* and guint8*.

**Comment by jaykrell** [comment]
Fyi, some of the operators come into play when you expand this to hashtable lookup:

```
custom-attrs.c:2578:86: error: invalid operands to binary expression
      ('monoeg_g_cast' and 'long')
  ...g_hash_table_lookup (image->weak_field_indexes, GINT_TO_POINTER (field_idx)) != NULL;
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~
```

"NULL" is of type long.

Longer term, the hashtables can templatized and only return an exact more derived type, and then no need for casting.

**Comment by jaykrell** [comment]
Also limit of gcc4.4/Cento6 is quick to reach here.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build OS X x64

**Comment by jaykrell** [comment]
Should be better now.

This stuff is a bit touchy, in that if you extend it to hashtable lookup, you need more functions, like compare to long. And if you add much, you exceed g++4.4 C++11 support.

But this should be ok.

The rvalue constructor can be removed if the two ternary operators out in the tree are tweaked, to cast or not use ternary.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT+LLVM

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Acceptance Tests

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build OS X x64

**Comment by jaykrell** [comment]
@monojenkins build OS X x64

**Comment by lambdageek** [comment]
I like this version of `g_cast_t<>` much better.  it does a lot fewer things.

File: mono/eglib/glib.h [hunk 13]
```diff
@@ -54,6 +54,122 @@
...
+// g_cast converts void* to T*.
+// e.g. #define malloc(x) (g_cast (malloc (x)))
+// FIXME It used to do more. Rename?
+struct g_cast
+{
+private:
+	void * const x;
+public:
+	explicit g_cast (void *y) : x(y) { }
+	g_cast (g_cast&& y) : x(y.x) { }
```
Comment (jaykrell) [inline mono/eglib/glib.h]: This line can be changed to the mildly preferred `= delete` if two occurences of ternary operator in the tree gain a cast or lose the ternary operator.

Comment (lambdageek) [reply mono/eglib/glib.h]: Let's fix the ternaries and put `= delete` here.

File: mono/mini/mini-amd64.c (hunk 12 above)
Comment (lambdageek) [reply mono/mini/mini-amd64.c]: This function is also defined in `mini-sparc.c`, `mini-s390x.c`, `mini-mips.c`, etc.

My preference would be not to fixup `mono_arch_instrument_prolog`/epilog in this PR.  Let's just get `g_cast_t<>` merged in the cases where we need it.

File: mono/eglib/gmodule.h [hunk 14]
```diff
@@ -20,6 +20,9 @@ typedef enum {
 	G_MODULE_BIND_LOCAL = 0x02,
 	G_MODULE_BIND_MASK = 0x03
 } GModuleFlags;
+
+G_ENUM_FUNCTIONS (GModuleFlags)
```
Comment (lambdageek) [inline mono/eglib/gmodule.h]: I wonder if each of the uses of `G_ENUM_FUNCTIONS` should come with a "fixme" note.

**Comment by jaykrell** [comment]
One subtlety I missed that Aleksey pointed out, is "which operators?".
Windows.h doesn't provide +/-/++, "only bit stuff".
Though there is the a & (a - 1) idiom like to see if only one bit set or something.

Maybe we have too many here. Maybe it doesn't matter.
In C all the operators are available. We use +/- for wait results I think, and ++ for stop-the-world phases. Though could get casts though, if we want fewer operators.

Everything can get casts, really.
Also ints might be appropriate.
Though, I an see where this goes -- ints are too weak. You want your own little ranges/units of numbers for different things, and don't accidentally mix them. Inches can be added to feet, but not to square inches, etc. stop-the-world phase is different than a wait result.

**Comment by jaykrell** [comment]
See https://github.com/mono/mono/commit/3a3607faa8827a2ac994cd7385508b2beb10f11a for some of the flip side of the enum operators.

File: mono/eglib/glib.h [hunk 15]
```diff
@@ -518,6 +554,7 @@ void    g_array_set_size          (GArray *array, gint length);
 #define g_array_append_val(a,v)   (g_array_append_vals((a),&(v),1))
 #define g_array_insert_val(a,i,v) (g_array_insert_vals((a),(i),&(v),1))
 #define g_array_index(a,t,i)      *(t*)(((a)->data) + sizeof(t) * (i))
+//FIXME previous missing parens
```
Comment (luhenry) [inline mono/eglib/glib.h]: Remove that.

File: mono/eglib/glib.h [hunk 16]
```diff
@@ -549,6 +586,7 @@ gpointer  *g_ptr_array_free               (GPtrArray *array, gboolean free_seg);
 void       g_ptr_array_foreach            (GPtrArray *array, GFunc func, gpointer user_data);
 guint      g_ptr_array_capacity           (GPtrArray *array);
 #define    g_ptr_array_index(array,index) (array)->pdata[(index)]
+//FIXME previous missing parens
```
Comment (luhenry) [inline mono/eglib/glib.h]: Remove that.

File: mono/eglib/glib.h [hunk 17]
```diff
@@ -1148,4 +1186,22 @@ glong     g_utf8_pointer_to_offset (const gchar *str, const gchar *pos);
...
+// Macros do not recurse, so naming function and macro the same is ok.
+// However these are also already macros.
+#undef g_malloc
+#undef g_realloc
+#undef g_malloc0
+#undef g_calloc
+#undef g_try_malloc
+#undef g_try_realloc
+#undef g_memdup
+#define g_malloc(x) (g_cast (monoeg_malloc (x)))
```
Comment (luhenry) [inline mono/eglib/glib.h]: Just declare that at https://github.com/mono/mono/pull/10027/files#diff-b8c6c2ba344450884373fb00f33685adR186 instead, and rename the `g_malloc` function to `g_malloc_internal` if it collides.

File: mono/metadata/handle.h [hunk 18]
```diff
@@ -245,7 +245,7 @@ Icall macros
 	do {							\
 		void* __result = MONO_HANDLE_RAW (HANDLE);	\
 		CLEAR_ICALL_FRAME;				\
-		return __result;				\
+		return g_cast (__result);			\
```
Comment (lambdageek) [inline mono/metadata/handle.h]: @luhenry this use of `g_cast` in `HANDLE_FUNCTION_RETURN_OBJ` is difficult to avoid unless we change the macro to take a return type.

Instead of this:

```c
MONO_RT_EXTERNAL_ONLY
MonoString *
some_api_function (const char *s)
{
  HANDLE_FUNCTION_ENTER();
  MonoStringHandle h = mono_string_new_handle (s);
  HANDLE_FUNCTION_RETURN_OBJ (h);
}
```

We would have to write

```c
MonoString *
some_api_function (const char *s)
{
  HANDLE_FUNCTION_ENTER ();
  MonoStringHandle h = mono_string_new_handle (s);
  HANDLE_FUNCTION_RETURN_OBJ (MonoString, h);
}
```

File: mono/eglib/glib.h (hunk 17 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: And then breakpoints on `g_malloc` no longer work.

File: mono/metadata/handle.h (hunk 18 above)
Comment (jaykrell) [reply mono/metadata/handle.h]: This one can use `auto` instead. Two of the handle.h changes can be `auto`, the other one I didn't look at closely. Because the type usually matches from `MONO_HANDLE_RAW`.

**Comment by jaykrell** [comment]
I'm torn on autocast or not. I originally had everything working with explicit casts. But there were many and it bothered people.

File: mono/eglib/glib.h (hunk 17 above)
Comment (jaykrell) [reply mono/eglib/glib.h]: Er, breakpoint is `monoeg_malloc`. I guess just break on `malloc`, though sometimes the extra layer is nice, i.e. to break only on mono's calls.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 FullAOT
